lib/spack/spack/spec.py

```diff
@@ -3514,76 +3656,72 @@ def satisfies(self, other, deps=True, strict=False):
                 return False
-        elif strict and (other.compiler and not self.compiler):
+        elif other.compiler and not self.compiler:
             return False
 
-        var_strict = strict
-        if (not self.name) or (not other.name):
-            var_strict = True
-        if not self.variants.satisfies(other.variants, strict=var_strict):
+        if not self.variants.satisfies(other.variants):
             return False
 
         # Architecture satisfaction is currently just string equality.
         # If not strict, None means unconstrained.
         if self.architecture and other.architecture:
-            if not self.architecture.satisfies(other.architecture, strict):
+            if not self.architecture.satisfies(other.architecture):
                 return False
-        elif strict and (other.architecture and not self.architecture):
+        elif other.architecture and not self.architecture:
             return False
 
-        if not self.compiler_flags.satisfies(other.compiler_flags, strict=strict):
+        if not self.compiler_flags.satisfies(other.compiler_flags):
             return False
 
         # If we need to descend into dependencies, do it, otherwise we're done.
         if deps:
-            deps_strict = strict
-            if self._concrete and not other.name:
-                # We're dealing with existing specs
-                deps_strict = True
-            return self.satisfies_dependencies(other, strict=deps_strict)
+            return self._satisfies_dependencies(other)
         else:
             return True
 
-    def satisfies_dependencies(self, other, strict=False):
-        """
-        This checks constraints on common dependencies against each other.
-        """
+    def _satisfies_dependencies(self, other):
         other = self._autospec(other)
 
         # If there are no constraints to satisfy, we're done.
         if not other._dependencies:
             return True
 
-        if strict:
-            # if we have no dependencies, we can't satisfy any constraints.
-            if not self._dependencies:
-                return False
-
-            # use list to prevent double-iteration
-            selfdeps = list(self.traverse(root=False))
-            otherdeps = list(other.traverse(root=False))
-            if not all(any(d.satisfies(dep, strict=True) for d in selfdeps) for dep in otherdeps):
-                return False
+        # If we have no dependencies, we can't satisfy any constraints.
+        if not self._dependencies:
+            return False
 
-        elif not self._dependencies:
-            # if not strict, this spec *could* eventually satisfy the
-            # constraints on other.
-            return True
+        # use list to prevent double-iteration
+        selfdeps = list(self.traverse(root=False))
+        otherdeps = list(other.traverse(root=False))
+        if not all(any(d.satisfies(dep) for d in selfdeps) for dep in otherdeps):
```

Comment:
Somewhere here we there should be a comment that for abstract specs the graph structure is more or less ignored (direct deps of abstract specs may concretize to deps at any depth...). As far as I understand, we only ever hit this when the rhs is abstract (`concrete.satisfies(abstract)`). Documenting these invariants makes it a bit easier to follow.

Reply:
> As far as I understand, we only ever hit this when the rhs is abstract

Yes, since we have the fast way out:
```python
if other.concrete:
    # The left-hand side must be the same singleton with identical hash. Notice that
    # package hashes can be different for otherwise indistinguishable concrete Spec
    # objects.
    return self.concrete and self.dag_hash() == other.dag_hash()
```
I think we can possibly remove some redundancy between here and the lines below:
```python
for name in self.common_dependencies(other):
    if not self[name].satisfies(other[name], deps=False):
        return False
```
but I'll try to explore that in future PRs.

Reply:
See 00f68d94d8793e3a8e16680fa40dfb0b87e3934b

Reply:
Removed redundancy in cf3a79a2e98a9881382a4d898125d5e9389ef9c0 - so I think this one is done too.